### DISCUSSION TITLE
How do I make mikro-orm work in NextJS?

### DISCUSSION THREAD

**Discussion by daviddavid** [mikro-orm/mikro-orm#3907] [Q&A]
I've been trying to wrap my head around Mikro-ORM, reading the docs, issues, etc. but keep running into roadblocks, so bear with me...

I forked the example [mikro-orm-nextjs](https://github.com/daviddavid/mikro-orm-nextjs) repo (updated mikro-orm to 5.6.3 and replaced sqlite with postgres) for a simple demo. It contains two entities, User and Books. Each User can have multiple Books. I want to create a User, a Book, and then add the book to the user. I created the following two APIs to demonstrate the process:

```Javascript
// /api/createUser

const em = getEM();

const user = new User('Foo Bar', 'foo@bar.com');
await em.persistAndFlush(user);
```

```Javascript
// /api/createBook

const em = getEM();

const user = await em.findOne(User, { id: 1 });

const book = new Book();
book.title = 'Book Title';

await em.persist(book);

user.books.add(book);
await em.flush();
```

Calling `/api/createUser` creates a user, as expected. However, calling `/api/createBook` afterward, throws the following error:

```
error - ValidationError: Trying to persist not discovered entity of type Book. Entity with this name was discovered, but not the prototype you are passing to the ORM. If using EntitySchema, be sure to point to the implementation via `class`.
```

Curiously, if you restart the server in between calls, `/api/createBook` works. However, (after having created a User) calling `/api/createBook` works, but then `/api/createUser` throws the same error:

```
error - ValidationError: Trying to persist not discovered entity of type User. Entity with this name was discovered, but not the prototype you are passing to the ORM. If using EntitySchema, be sure to point to the implementation via `class`.
```

So, what am I doing wrong? These errors don't tell me anything, when the same calls work after server restarts :(

**Comment by B4nan** [comment]
Maybe try asking the author of that example. The integration looks a bit hackish to me but I never used nextjs myself, so I cant really help - not without deeper experiments with it - and I don't have the time for that right now. 

Not sure if there are some kind of middlewares (express) or hooks (fastify) in nextjs? As that is the suggested way of registering the request context.

**Reply by B4nan** [reply]
Did a quick search and here is some example of using inversify with nextjs, that would be clearner way to do things dependency-wise:

https://featuringcode.com/oop-and-dependency-injection-in-nextjs

You can see how to register the ORM services to inversify in another example repo (although this is one quite complex):

https://github.com/PodaruDragos/inversify-example-app

I guess it would deserve proper integration with a guide, I was hoping for some nextjs users to provide it, instead of me, who never used nextjs :D

**Reply by B4nan** [reply]
Here is some other example app, much more complex, apparently uses similar approach (but not exactly the same methods are used).

https://github.com/Arslanoov/red

**Reply by B4nan** [reply]
This might be also related

https://github.com/mikro-orm/mikro-orm/issues/3167

**Comment by daviddavid** [comment]
Thanks for the suggestions. Turns out there already is a ticket surrounding the issue: https://github.com/jonahallibone/mikro-orm-nextjs/issues/12

I was hoping it would be a simple fix or someone who got it to work could chime in. I'm not dead set on getting it to work in NextJS, so I'll probably just try it out in Express and see how it goes.

**Reply by octet-stream** [reply]
I think I had this issue before. I ended up just avoiding usage of entity classes directly. Try replace `const book = new Book()` with `const = em.create(Book, {})`. I made this change in your example and this does seem to solve your problem. Also, try to avoid `instanceof` usage on your entities - it does not work either :)

----

I noticed that you starred my [`next-mikro-orm-trpc-template`](https://github.com/octet-stream/next-mikro-orm-trpc-template) template few days ago, hope it helped to fix the issue. :)

**Reply by B4nan** [reply]
FYI `em.create()` gets entity metadata (by its string name) and uses the class reference that gets discovered, so that could help with some issues. But its shoudnt be needed, you should ideally discover the correct references, this is the reason why we have both `entities` and `entitiesTs` - its weird this would surface without folder based discovery.

Looked at the template and found two small things:
- your [runIsolatied](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/src/server/lib/db.ts#L96) helper is doing pretty much the same as [disableIdentityMap: true](https://mikro-orm.io/docs/next/entity-manager#disabling-identity-map-and-change-set-tracking)
- [`micro` vs `mikro`](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/package.json#L2) :] ([same here](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/package.json#L6))

I'd be curious if you can try [v6](https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/upgrading-v5-to-v6.md), as one of the already implemented changes are removal of static requires (#3814) which should generally help with projects that do bundling.

**Reply by octet-stream** [reply]
Tried with v6: the issue is still here. I assume that it happens on the Next.js end. Maybe it creates few entity classes for the same thing when bundling, I have no idea. But it appears to me, MikroORM and that [procedure](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/c049071863646313a625abfffc9c06a3768b3c81/src/server/trpc/route/pony/create.ts) have references to different objects (this is why `instanceof` check also returns `false`).

If you run the app, open `/new` page, then fill and send form - it will create a new row on the first run. Then, if you navigate to `/new` page again and try to create another record - you will get the same error message as mentioned in this topic. The error happens [here](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/c049071863646313a625abfffc9c06a3768b3c81/src/server/trpc/route/pony/create.ts#L16).

> [micro vs mikro](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/package.json#L2) :] ([same here](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/package.json#L6))

Oops, sorry, my mistake :D

> your [runIsolatied](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/master/src/server/lib/db.ts#L96) helper is doing pretty much the same as [disableIdentityMap: true](https://mikro-orm.io/docs/next/entity-manager#disabling-identity-map-and-change-set-tracking)

Not sure if that replaces `runIsolated` - it was made to run queries outside `RequestContext`, which happens rarely, but sometimes I need that. It essentially helps me to keep my code with DRY principe. I might also need to persist and flush entities inside of `runIsolated`, which should be possible, because I have `EntityManager` inside of the callback.
But thanks for the tip! I think in [this](https://github.com/octet-stream/eri/blob/bbe15dec8710570ec676ed4517dfc45d332d5583/src/pages/api/auth/%5B...nextauth%5D.ts#L107-L111) case it might replace `runIsolated`, because I only need to retrieve the data for user, and `next-auth` keeps ignoring `RequestContext`, even if I'd add a [middleware](https://github.com/octet-stream/eri/blob/bbe15dec8710570ec676ed4517dfc45d332d5583/src/server/middleware/withORMContext.ts) for that (By the way it could be done with [`next-connect`](https://www.npmjs.com/package/next-connect), I think you were asking about that in one of your comments before).

**Reply by octet-stream** [reply]
Another issue I personally have using MikroORM with Next.js is that it [can't discover enums](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/ac0cf4b2b6abd20a94addcd5379f91635359b4e5/src/server/db/entity/Pony.ts#L20-L21), if they were defined in [separate file](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/ac0cf4b2b6abd20a94addcd5379f91635359b4e5/src/server/trpc/type/common/PonyRace.ts#L3-L8). I did this because Next.js app will crash otherwise if I'll try to import the enum from entity module, because it will also import server-specific code on the client-side. The [documentation](https://mikro-orm.io/docs/defining-entities#enums) says I have to re-export the enum from entity module, but it does not help to solve the problem. Not sure if it supposed to, if I [don't use paths](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/ac0cf4b2b6abd20a94addcd5379f91635359b4e5/src/server/lib/db/config.ts#L17) to discover entities. Anyway, I fixed this by [extracting values](https://github.com/octet-stream/next-mikro-orm-trpc-template/blob/ac0cf4b2b6abd20a94addcd5379f91635359b4e5/src/server/db/entity/Pony.ts#L10) from the enum.

**Reply by B4nan** [reply]
The re-export suggestion is only for a deprecated use case, where you use ts-morph, the ORM detects the enum name and tries to import it dynamically (from the same file where the entity is defined). This is removed from docs and discouraged nowadays, you should provide the enum reference explicitly:

```ts
@Enum(() => PonyRace)
race!: PonyRace
```

Like that it does not work for you? I'll try to find some time to look at your project, but I am kinda swamped these days.

**Reply by octet-stream** [reply]
Nope, and it was the first thing I've tried following the documentation. This is the error I have:

<details>
<summary>Details</summary>

<img width="999" alt="image" src="https://user-images.githubusercontent.com/7884558/215129342-3e6f743a-fed2-4a79-97c1-4db4ad652c0b.png">

</details>

It happens because it discovers the type as `undefined` (you can see it in `MetadataDiscovery.getMappedType`), hence the error :)

**Reply by octet-stream** [reply]
> I'll try to find some time to look at your project, but I am kinda swamped these days.

Sure, feel free to ask me if you'll need my help regarding my template or Next.js
